File: android/src/main/kotlin/com/medallia/medallia_dxa/MedalliaDxaPlugin.kt
package com.medallia.medallia_dxa

import PerformanceData
import SamplingFlutterInfo
import SdkConfigInfo
import SessionInfo
import android.app.Activity
import android.os.SystemClock
import android.util.Log
import com.medallia.dxa.DXA
import com.medallia.dxa.buildercommon.MedalliaDXA
import com.medallia.dxa.common.entity.DXACollection
import com.medallia.dxa.common.entity.TouchEvent
import com.medallia.dxa.common.enums.CollectionDataType
import com.medallia.dxa.common.enums.CustomerConsentType
import com.medallia.dxa.common.enums.DeviceUnderStress
import com.medallia.dxa.common.enums.PlatformType
import com.medallia.dxa.common.internal.logic.providers.ActivityProvider
import com.medallia.dxa.common.internal.logic.providers.ActivityResumedListener
import com.medallia.dxa.common.internal.models.DXAConfig
import com.medallia.dxa.common.internal.models.ImageQualityLevel
import com.medallia.dxa.common.internal.models.Multiplatform
import com.medallia.dxa.common.internal.models.Performance
import com.medallia.dxa.common.internal.models.SdkConfig
import com.medallia.dxa.common.internal.models.Session
import com.medallia.dxa.common.internal.models.SamplingMultiplatform
import com.medallia.medallia_dxa.Messages.FlutterError
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.android.FlutterFragmentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collect
import java.lang.ref.WeakReference
import java.util.Date

/** MedalliaDxaPlugin */
class MedalliaDxaPlugin : FlutterPlugin, Messages.MedalliaDxaNativeApiPigeon {

    private lateinit var medalliaDxa: MedalliaDXA

    private val logTag = "DXA-FLUTTER"

    private val enableLogs = false

    private var latestFlutterActivity: WeakReference<Activity> = WeakReference(null)

    private val binderScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    private var flutterEventSink: EventChannel.EventSink? = null

    private var initialized = false

    private var lastConfig: SdkConfig? = null

    private var minimumSupportedApi: Int = 26

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        log("Attaching to engine...")
        
        Messages.MedalliaDxaNativeApiPigeon.setUp(flutterPluginBinding.binaryMessenger, this)
        ActivityProvider.addListen(onFlutterActivityResumedListener)
        EventChannel(flutterPluginBinding.binaryMessenger, "medallia.dxa.streamChannel")
            .setStreamHandler(onFlutterStreamChannelListener)
        if(isApiSupported().not()){ return }
        medalliaDxa = DXA.getInstance(flutterPluginBinding.applicationContext)
        startCollectSdkConfig()
        startCollectPerformanceData()
        startCollectingSamplingData()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        log("Detaching from engine...")
        Messages.MedalliaDxaNativeApiPigeon.setUp(binding.binaryMessenger, null)
        ActivityProvider.removeListener(onFlutterActivityResumedListener)
        latestFlutterActivity.clear()
        binderScope.cancel()
    }

    override fun initialize(
        msg: Messages.SessionMessagePigeon,
        result: Messages.Result<Messages.LiveConfigurationFromInitializePigeon>
    ) {
        if(isApiSupported().not()){
            val liveConfig = liveConfigForBlockedSdk(msg.version)
            return result.success(liveConfig)
        }

        val initTime = Date().time
        log(
            message = "calling initialize: account=${msg.account} - property${msg.property} - " +
                    "version=${msg.version} - consents=${msg.consents}"
        )
        val consents = translateConsentsFlutterToAndroid(msg.consents)

        binderScope.launch {
            val config: SdkConfig?
            if (initialized) {
                log(message = "already initialized")
                config = lastConfig
            } else {
                config = medalliaDxa.standaloneInitialize(
                    dxaConfig = DXAConfig(
                        customerConsent = consents,
                        accountId = msg.account,
                        propertyId = msg.property,
                        mobileDataEnabled = msg.mobileDataEnabled,
                        enhancedLogsEnabled = msg.enhancedLogs
                    ),
                    platform = Multiplatform(
                        type = PlatformType.FLUTTER,
                        version = msg.version
                    )
                )
                initialized = true
                lastConfig = config
            }

            log(message = "Initial config: $config")
            val endTime = Date().time
            log(message = "initialize SDK took: ${endTime - initTime}")
            val liveConfiguration = Messages.LiveConfigurationFromInitializePigeon()
            liveConfiguration.overrideUserConfig = config?.mscOverrideUserConfigs
            liveConfiguration.blockedFlutterAppVersions = config?.vcBlockedFlutterAppVersions
            liveConfiguration.blockedFlutterSDKVersions = config?.vcBlockedFlutterSDKVersions
            liveConfiguration.blockedNativeSDKVersions = config?.vcBlockedAndroidSDKVersions
            liveConfiguration.overrideUserConfig = config?.mscOverrideUserConfigs
            liveConfiguration.maskingColor = config?.mMaskingColor?.let(Integer::toHexString)
            liveConfiguration.imageQualityType = config?.vsImageQuality?.level?.toLong()
            liveConfiguration.videoQualityType = config?.vsVideoQuality?.level?.toLong()
            liveConfiguration.showLocalLogs = config?.daShowLocalLogs
            liveConfiguration.allowLocalLogs = config?.daAllowLogs
            liveConfiguration.disableScreenTracking = config?.dstDisableScreenTracking
            liveConfiguration.screensMasking = config?.dstScreenMasking
            liveConfiguration.appVersion = medalliaDxa.appVersion()
            liveConfiguration.nativeSDKVersion = medalliaDxa.sdkVersion()
            result.success(liveConfiguration)
        }
    }

    override fun startScreen(msg: Messages.StartScreenMessagePigeon, result: Messages.VoidResult) {
        msg.run {
            val initTime = Date().time
            log(
                message = "calling startScreen: screenID=${screenId} - screenName${screenName} - " +
                        "startTime=${startTime} - isBackground=${isBackground}"
            )
            val currentActivity = latestFlutterActivity.get() ?: ActivityProvider.currentActivity
            ?: let {
                return@run result.success()
            }
            val job =
                medalliaDxa.startScreen(
                    msg.screenId,
                    msg.screenName,
                    msg.startTime,
                    currentActivity
                )
            binderScope.launch {
                job.join()
                val endTime = Date().time
                log(message = "startScreen took: ${endTime - initTime}")
                result.success()
            }
        }
    }

    override fun endScreen(msg: Messages.EndScreenMessagePigeon, result: Messages.VoidResult) {
        msg.run {
            val initTime = Date().time
            log(
                message = "calling endScreen: screenID=${screenId} - screenName${screenName} - " +
                        "endTime=${endTime} - isBackground=${isBackground}"
            )
            val currentActivity = latestFlutterActivity.get() ?: ActivityProvider.currentActivity
            if (currentActivity == null) {
                result.success()
                return
            }
            val job =
                medalliaDxa.endScreen(
                    msg.screenId,
                    msg.screenName,
                    msg.endTime,
                    msg.screenRecordingEndTime,
                    currentActivity
                )
            binderScope.launch {
                job.join()
                val endTime = Date().time
                log(message = "endScreen took: ${endTime - initTime}")
                result.success()
            }
        }
    }

    override fun setConsents(consentLevel: Long) {
        medalliaDxa.setConsent(
            translateConsentsFlutterToAndroid(consentLevel)
        )
    }

    override fun saveScreenshot(msg: Messages.ScreenshotMessagePigeon, result: Messages.VoidResult) {
        msg.run {
            val initTime = Date().time
            log(
                message = "calling saveScreenshot: screenID=${screenId} - screenName${screenName} - " +
                        "startFocusTime=${startFocusTime} - screenshotDataSize=${screenshotData.size} - " +
                        "isPlaceholder=${isPlaceholder}"
            )

            val job =
                medalliaDxa.saveScreenShot(screenshotData, screenId, screenName, startFocusTime, isPlaceholder)
            binderScope.launch {
                job.join()
                val endTime = Date().time
                log(message = "saveScreenshot took: ${endTime - initTime}")
                result.success()
            }
        }
    }

    override fun sendDimensionWithString(msg: Messages.DimensionStringMessagePigeon) {
        msg.let {
            medalliaDxa.sendCustomDimension(it.dimensionName, it.value)
        }
    }

    override fun sendDimensionWithNumber(msg: Messages.DimensionNumberMessagePigeon) {
        msg.let {
            medalliaDxa.sendCustomDimension(it.dimensionName, it.value)
        }
    }


    override fun sendDimensionWithBool(msg: Messages.DimensionBoolMessagePigeon) {
        msg.let {
            medalliaDxa.sendCustomDimension(it.dimensionName, it.value)
        }
    }

    override fun sendGoal(msg: Messages.GoalMessagePigeon) {
        if (msg.value == null) {
            medalliaDxa.sendGoal(msg.goal)
        } else {
            medalliaDxa.sendGoal(msg.goal, msg.value!!)
        }
    }

    override fun sendError(errorMessage: String) {
        medalliaDxa.sendError(errorMessage)
    }

    override fun sendDataOverWifiOnly(value: Boolean) {
        medalliaDxa.sendDataOverWifiOnly(enabled = value)
    }

    override fun sendHttpError(msg: Long) {
        msg.toInt().let(medalliaDxa::sendHttpError)
    }

    override fun sendImageQuality(imageQuality: Long) {

        val imageQualityLevel = when (imageQuality) {
            0L -> ImageQualityLevel.Poor
            1L -> ImageQualityLevel.Low
            2L -> ImageQualityLevel.Average
            3L -> ImageQualityLevel.High
            4L -> ImageQualityLevel.Ultra
            else -> {
                ImageQualityLevel.Average
            }
        }

        medalliaDxa.setImageQuality(imageQualityLevel)
    }


    override fun getWebViewProperties(result: Messages.Result<String>) {
        log(message = "calling getWebViewProperties")
        result.success(medalliaDxa.getWebViewParams())
    }

    override fun getSessionId(result: Messages.Result<String>) {
        log(message = "calling getSessionId")
        result.success(medalliaDxa.getSessionId())
    }

    override fun getSessionUrl(result: Messages.Result<String>) {
        log(message = "calling getSessionUrl")
        medalliaDxa.getSessionUrl {sessionUrl ->
            if (sessionUrl != null) {
                result.success(sessionUrl)
            } else{
                result.error(FlutterError("Session url", "Session url is null", ""))
            }
        }
    }

    override fun setRetention(msg: Boolean) {
        msg.let {
            medalliaDxa.setRetention(it)
        }
    }

    override fun saveLog(log: String) {
        log.let {
            medalliaDxa.saveLog(it)
        }
    }

    override fun sendTouchEvents(touchEvents: MutableList<Messages.TouchEventNativeClassPigeon>, result: Messages.VoidResult) {

        val dxaCollection = mutableListOf<DXACollection>()

        touchEvents.forEach { touchEvent ->
            val touchesMapped: List<TouchEvent> = touchEvent.d.mapNotNull { touchDetail ->
                val x = touchDetail?.x
                val y = touchDetail?.y
                if (x != null && y != null) {

                    TouchEvent(
                        x = x.toInt(),
                        y = y.toInt(),
                        clickForce = touchDetail.cf.toInt(),
                        offsetX = touchDetail.ox?.toInt(),
                        offsetY = touchDetail.oy?.toInt(),
                        scrollId = touchDetail.sid?.toInt(),
                        parentScrollId = touchDetail.pid?.toInt()
                    )
                } else {
                    null
                }
            }

            val dxaCollectionTouchEvent = DXACollection(
                System.nanoTime(),
                touchEvent.o.toInt().toLong(),
                translateTouchType(touchEvent.t),
                touchesMapped,
            )
            log(dxaCollectionTouchEvent.t)
            dxaCollection.add(dxaCollectionTouchEvent)
        }

        val job =  medalliaDxa.saveGestures(dxaCollection)
        binderScope.launch {
            job.join()
            result.success()
        }
    }

    private fun startCollectPerformanceData() {

        binderScope.launch {
            medalliaDxa.getDevicePerformanceFlow().collect { performance: Performance ->

                val cpuUsage = (performance.c / performance.nC) * 100
                val batteryLevel = performance.b * 100
                val memoryUsage = (performance.m / performance.mM) * 100
                val deviceUnderStress = when (performance.uS.lastOrNull()) {
                    DeviceUnderStress.HIGH_CPU_USAGE.key -> DeviceUnderStress.HIGH_CPU_USAGE
                    DeviceUnderStress.HIGH_MEMORY_USAGE.key -> DeviceUnderStress.HIGH_MEMORY_USAGE
                    DeviceUnderStress.LOW_BATTERY_LEVEL.key -> DeviceUnderStress.LOW_BATTERY_LEVEL
                    else -> DeviceUnderStress.NOT_STRESSED
                }

                log(
                    message = "new performance data: cpuUsage=$cpuUsage, batteryLevel=$batteryLevel, " +
                            "memoryUsage=$memoryUsage, deviceUnderStress=$deviceUnderStress",
                    level = Log.VERBOSE
                )

                val performanceData = PerformanceData(
                    cpuUsage = cpuUsage,
                    batteryLevel = batteryLevel,
                    memoryUsage = memoryUsage,
                    deviceUnderStress = deviceUnderStress
                )

                val newConfigMap = performanceData.toHashMap()

                newConfigMap["eventChannelId"] = "performance_metrics"
                flutterEventSink?.success(newConfigMap)
            }
        }
    }

    private fun startCollectSdkConfig() {
        binderScope.launch {
            medalliaDxa.getConfigFlow().collect { newConfig: SdkConfig? ->
                log(
                    message = "new SDK config from native. $newConfig",
                    level = Log.VERBOSE
                )
                lastConfig = newConfig
                if (newConfig == null) {
                    return@collect
                }
                val newConfigInfo = SdkConfigInfo(
                    mscOverrideUserConfigs = newConfig.mscOverrideUserConfigs,
                    vcBlockedFlutterSDKVersions = newConfig.vcBlockedFlutterSDKVersions,
                    vcBlockedFlutterAppVersions = newConfig.vcBlockedFlutterAppVersions,
                    vcBlockedNativeSDKVersions = newConfig.vcBlockedAndroidSDKVersions,
                    mMaskingColor = newConfig.mMaskingColor,
                    vsImageQuality = newConfig.vsImageQuality,
                    vsVideoQuality = newConfig.vsVideoQuality,
                    daShowLocalLogs = newConfig.daShowLocalLogs,
                    dstDisableScreenTracking = newConfig
                        .dstDisableScreenTracking,
                    daAllowLogs = newConfig.daAllowLogs,
                    dstScreenMasking = newConfig.dstScreenMasking
                )

                val newConfigMap = newConfigInfo.toHashMap()

                newConfigMap["eventChannelId"] = "live_configuration"
                flutterEventSink?.success(newConfigMap)

            }
        }
    }

    private fun startCollectingSamplingData() {
        binderScope.launch {
            medalliaDxa.getSampling().collect { samplingData: SamplingMultiplatform ->
                log(
                    message = "new sampling data from native. $samplingData",
                    level = Log.VERBOSE
                )
                val samplingFlutterInfo = SamplingFlutterInfo(
                    stopTrackingDueToSampling = samplingData.stopTrackingDueToSampling,
                    stopRecordingDueToSampling = samplingData.stopRecordingDueToSampling
                )
                val samplingFlutterMap = samplingFlutterInfo.toHashMap()

                samplingFlutterMap["eventChannelId"] = "sampling_data"
                flutterEventSink?.success(samplingFlutterMap)

            }
        }
    }

    private val onFlutterActivityResumedListener = object : ActivityResumedListener {
        override fun onActivityResumed(activity: Activity) {
            if (activity is FlutterActivity || activity is FlutterFragmentActivity) {
                log("Detected flutter activity: $activity", level = Log.INFO)
                latestFlutterActivity = WeakReference(activity)
            }
        }
    }

    private val onFlutterStreamChannelListener = object : EventChannel.StreamHandler {

        override fun onListen(
            arguments: Any?, eventSink: EventChannel.EventSink?
        ) {
            flutterEventSink = eventSink
        }

        override fun onCancel(arguments: Any?) {
            flutterEventSink = null
        }
    }

    private fun translateConsentsFlutterToAndroid(consents: Long): CustomerConsentType {
        return when (consents) {
            1L -> CustomerConsentType.ANALYTICS
            2L -> CustomerConsentType.ANALYTICS_AND_RECORDING
            else -> CustomerConsentType.NONE
        }
    }

    private fun translateTouchType(value: Long): String {
        return when (value) {
            0L -> CollectionDataType.TOUCH_BEGIN.value
            1L -> CollectionDataType.TOUCH_MOVE.value
            2L -> CollectionDataType.TOUCH_END.value
            else -> CollectionDataType.TOUCH_END.value
        }
    }

    private fun log(message: String, level: Int = Log.DEBUG) {
        if (!enableLogs) return
        when (level) {
            Log.VERBOSE -> Log.v(logTag, message)
            Log.DEBUG -> Log.d(logTag, message)
            Log.INFO -> Log.i(logTag, message)
            Log.ERROR -> Log.e(logTag, message)
            Log.WARN -> Log.w(logTag, message)
        }
    }

    private fun isApiSupported(): Boolean {
        return android.os.Build.VERSION.SDK_INT >= minimumSupportedApi
    }

    private fun liveConfigForBlockedSdk(sdkVersion: String): Messages.LiveConfigurationFromInitializePigeon {
        val fakeLiveConfig = Messages.LiveConfigurationFromInitializePigeon()
        fakeLiveConfig.blockedFlutterSDKVersions = listOf(sdkVersion)
        return fakeLiveConfig
    }

}

File: ios/medallia_dxa/Sources/medallia_dxa/MedalliaDxaPlugin.swift
import Flutter
import MedalliaDXAFlutter
import UIKit

public class MedalliaDxaPlugin: NSObject, FlutterPlugin, MedalliaDxaNativeApiPigeon {

    private static var flutterEventSink: FlutterEventSink? = nil
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger: FlutterBinaryMessenger = registrar.messenger()
        let api: MedalliaDxaNativeApiPigeon & NSObjectProtocol = MedalliaDxaPlugin.init()
        MedalliaDxaNativeApiPigeonSetup.setUp(binaryMessenger: messenger, api: api)
        FlutterEventChannel(name: "medallia.dxa.streamChannel", binaryMessenger: messenger)
            .setStreamHandler(onPerformanceChangeListener)
    }
    func initialize(
        msg: SessionMessagePigeon,
        completion: @escaping (Result<LiveConfigurationFromInitializePigeon, any Error>) -> Void
    ) {
        var liveConfiguration: MedalliaDXAFlutter.LiveConfigurationMultiplatform

        let nativeConsents: Consent = translateConsentsFlutterToIos(
            flutterConsents: Int(msg.consents))
        let configuration = Configuration(
            account: String(describing: msg.account), property: String(describing: msg.property),
            consent: nativeConsents, crashReporterEnabled: msg.crashReporterEnabled,
            mobileDataEnable: msg.mobileDataEnabled, enhancedLogsEnabled: msg.enhancedLogs)
        liveConfiguration = DXA.initialize(
            configuration: configuration,
            multiplatform: Platform(
                type: .flutter, version: String(describing: msg.version), language: "Dart"),
            dxaDelegate: self)

        let fltLiveConfiguration = buildLiveConfigurationFromInitializeClass(
            liveConfiguration: liveConfiguration)
        completion(.success(fltLiveConfiguration))
    }

    func startScreen(
        msg: StartScreenMessagePigeon, completion: @escaping (Result<Void, any Error>) -> Void
    ) {
        DXA.startNewScreen(name: msg.screenName, id: Int(msg.screenId), timestampInMilliseconds: Int(msg.startTime))
        completion(.success(()))
    }

    func endScreen(msg: EndScreenMessagePigeon, completion: @escaping (Result<Void, any Error>) -> Void) {
        DXA.endScreen(endRecordingTime: TimeInterval(msg.endTime / 1000))
        completion(.success(()))
    }

    func setConsents(value: Int64) throws {
        let nativeConsents: Consent = translateConsentsFlutterToIos(flutterConsents: Int(value))
        DXA.setConsent(nativeConsents)
    }

    func saveScreenshot(
        msg: ScreenshotMessagePigeon, completion: @escaping (Result<Void, any Error>) -> Void
    ) {
        DXA.saveScreenShot(screenshot: msg.screenshotData.data, isPlaceholder: msg.isPlaceholder, timestampInMilliseconds: Int(msg.startFocusTime))
        completion(.success(()))
    }

    func sendTouchEvents(touchEvents: [TouchEventNativeClassPigeon], completion: @escaping (Result<Void, any Error>) -> Void) {
        let flutterEvents: [FlutterEvent] = touchEvents.compactMap { touchEvent -> FlutterEvent in
            let touchesMapped: [FlutterTouch] = touchEvent.d.compactMap { touchDetail in
                
                guard let x = touchDetail?.x, let y = touchDetail?.y
                else {
                    return nil
                }
                return FlutterTouch(
                    x: Int(x),
                    y: Int(y),
                    offsetX: convertInt64ToInt(touchDetail?.oX),
                    offsetY: convertInt64ToInt(touchDetail?.oY),
                    clickForce: convertInt64ToInt(touchDetail?.cf),
                    scrollID: convertInt64ToInt(touchDetail?.sID),
                    parentScrollID: convertInt64ToInt(touchDetail?.pID)
                )
            }
            return FlutterEvent(
                offset: Int(touchEvent.o),
                data: touchesMapped,
                type: translateTouchType(value: touchEvent.t)
            )
        }
        DXA.sendTouchEvents(touches: flutterEvents)
        completion(.success(()))
    }

    func sendDimensionWithString(msg: DimensionStringMessagePigeon) throws {
        DXA.send(dimension: msg.dimensionName, value: msg.value)
    }

    func sendDimensionWithNumber(msg: DimensionNumberMessagePigeon) throws {
        DXA.send(dimension: msg.dimensionName, value: msg.value)
    }

    func sendDimensionWithBool(msg: DimensionBoolMessagePigeon) throws {
        DXA.send(dimension: msg.dimensionName, value: msg.value)
    }

    func sendGoal(msg: GoalMessagePigeon) throws {

        guard let goalValue = msg.value else {
            DXA.send(goal: msg.goal)
            return
        }
        DXA.send(goal: msg.goal, with: Float(goalValue))

    }

    func sendError(errorMessage: String) throws {
        DXA.sendError(errorMessage)
    }

    func sendDataOverWifiOnly(value: Bool) throws {
        DXA.mobileDataEnable = !value
    }

    func sendHttpError(msg: Int64) throws {
        DXA.sendHTTPError(statusCode: Int(msg))
    }

    func sendImageQuality(imageQuality: Int64) throws {
        guard let imageQualityValue = ImageQualityType(rawValue: Int(imageQuality)) else {
            return
        }
        DXA.setImageQuality(imageQualityValue)
    }

    func setRetention(value: Bool) throws {
        DXA.setRetention(value)

    }

    func saveLog(log: String) throws {
        DXA.saveLog(message: log)
    }

    func getWebViewProperties(completion: @escaping (Result<String, any Error>) -> Void) {
        let webViewProperties = DXA.webViewProperties
        if webViewProperties != nil {
            completion(.success(webViewProperties!))
            return
        }

        completion(
            .failure(
                PigeonError(
                    code: "getWebViewProperties",
                    message: "Unexpected null value, session has not been initialized", details: "")
            ))

    }

    func getSessionId(completion: @escaping (Result<String, any Error>) -> Void) {
        let sessionId = DXA.sessionId
        if sessionId != nil {
            completion(.success(sessionId!))
            return
        }

        completion(
            .failure(
                PigeonError(
                    code: "getSessionId",
                    message: "Unexpected null value, session has not been initialized", details: nil
                )))

    }

    func getSessionUrl(completion: @escaping (Result<String, any Error>) -> Void) {
        //Check to prevent exceptions when multiple sessionURLs are returned
        var hasCompletionBeenCalled = false
        DXA.sessionURL = { sessionUrl in
            if hasCompletionBeenCalled {
                return
            }
            hasCompletionBeenCalled = true
            if sessionUrl != nil {
                completion(.success(sessionUrl!))
                return
            }
            completion(
                .failure(
                    PigeonError(
                        code: "getSessionUrl",
                        message: "Unexpected null value, session has not been initialized",
                        details: nil)))
        }
    }

    private func translateConsentsFlutterToIos(flutterConsents value: Int) -> Consent {
        var nativeConsent: Consent

        switch value {
        case 0:
            nativeConsent = .none
        case 1:
            nativeConsent = .analytics
        case 2:
            nativeConsent = .analyticsAndRecording
        default:
            nativeConsent = .none
        }
        return nativeConsent

    }

    private static var onPerformanceChangeListener: NSObjectProtocol & FlutterStreamHandler = {
        class PerformanceChangeListener: NSObject, FlutterStreamHandler {

            func onListen(
                withArguments arguments: Any?,
                eventSink: @escaping FlutterEventSink
            ) -> FlutterError? {
                flutterEventSink = eventSink
                return nil
            }

            func onCancel(withArguments arguments: Any?) -> FlutterError? {
                flutterEventSink = nil
                return nil
            }
        }

        return PerformanceChangeListener()
    }()

    private func buildLiveConfigurationFromInitializeClass(
        liveConfiguration: MedalliaDXAFlutter.LiveConfigurationMultiplatform
    ) -> LiveConfigurationFromInitializePigeon {

        var imageQualityType: Int64? = nil
        if liveConfiguration.imageQualityType != nil {
            imageQualityType = Int64(liveConfiguration.imageQualityType!)
        }

        var videoQualityType: Int64? = nil
        if liveConfiguration.videoQualityType != nil {
            videoQualityType = Int64(liveConfiguration.videoQualityType!)
        }

        let fltLiveConfiguration = LiveConfigurationFromInitializePigeon(
            overrideUserConfig: liveConfiguration.useLiveConfiguration,
            blockedFlutterSDKVersions: liveConfiguration.blockedFlutterSDKVersions,
            blockedFlutterAppVersions: liveConfiguration.blockedFlutterAppVersions,
            blockedNativeSDKVersions: liveConfiguration.sdkVersionsBlocked,
            maskingColor: liveConfiguration.maskingColor,
            showLocalLogs: liveConfiguration.showLocalLogs,
            allowLocalLogs: liveConfiguration.allowLogs, 
            imageQualityType: imageQualityType,
            videoQualityType: videoQualityType,
            disableScreenTracking: liveConfiguration.disableScreenTracking,
            screensMasking: liveConfiguration.screensMasking,
            appVersion: DXA.appVersion,
            nativeSDKVersion: DXA.sdkVersion
        )
        return fltLiveConfiguration
    }

}
extension MedalliaDxaPlugin: DXADelegate {
    public func sampling(_ sampling: MedalliaDXAFlutter.SamplingMultiplatform) {

        if MedalliaDxaPlugin.flutterEventSink == nil { return }
        do {
            let samplingFlutterData = SamplingFlutterData(
                stopTrackingDueToSampling: sampling.stopTrackingDueToSampling,
                stopRecordingDueToSampling: sampling.stopRecordingDueToSampling
            )

            let jsonData = try JSONEncoder().encode(samplingFlutterData)
            var dict =
                try JSONSerialization.jsonObject(with: jsonData, options: .allowFragments)
                as? [String: Any]
            if dict == nil {
                throw NSError(
                    domain: "DictError", code: 0,
                    userInfo: [NSLocalizedDescriptionKey: "Dict is nil"])
            }
            dict?["eventChannelId"] = "sampling_data"
            DispatchQueue.main.async {
                MedalliaDxaPlugin.flutterEventSink?(dict)
            }
        } catch {
            print("Sampling data JSON serialization error: \(error)")
        }
    }

    public func liveConfig(_ configuration: MedalliaDXAFlutter.LiveConfigurationMultiplatform) {

        if MedalliaDxaPlugin.flutterEventSink == nil { return }
        do {

            let liveConfigurationFlutterData = LiveConfigurationFlutterData(
                useLiveConfiguration: configuration.useLiveConfiguration,
                disableScreenTracking: configuration.disableScreenTracking,
                screensMasking: configuration.screensMasking,
                imageQualityType: configuration.imageQualityType,
                videoQualityType: configuration.videoQualityType,
                maskingColor: configuration.maskingColor,
                showLocalLogs: configuration.showLocalLogs,
                allowLocalLogs: configuration.allowLogs,
                blockedFlutterSDKVersions: configuration.blockedFlutterSDKVersions,
                blockedFlutterAppVersions: configuration.blockedFlutterAppVersions,
                blockedNativeSDKVersions: configuration.sdkVersionsBlocked
            )

            let jsonData: Data = try JSONEncoder().encode(liveConfigurationFlutterData)
            var dict =
                try JSONSerialization.jsonObject(with: jsonData, options: .allowFragments)
                as? [String: Any]
            if dict == nil {
                throw NSError(
                    domain: "DictError", code: 0,
                    userInfo: [NSLocalizedDescriptionKey: "Dict is nil"])
            }
            dict?["eventChannelId"] = "live_configuration"
            DispatchQueue.main.async {
                MedalliaDxaPlugin.flutterEventSink?(dict)
            }
        } catch {
            print("Live Config JSON serialization error: \(error)")
        }
    }

    public func performance(_ data: PerformanceMultiplatform) {

        if MedalliaDxaPlugin.flutterEventSink == nil { return }
        do {
            var isStressedString: String
            if let firstStressedValue = data.isStressed.first {
                isStressedString = translateDeviceStressedTypeToString(
                    nativeStresstedType: firstStressedValue)
            } else {
                isStressedString = "none"
            }
            let performanceFlutterData = PerformanceFlutterData(
                isStressed: isStressedString,
                battery: data.battery,
                memory: data.memory,
                cpu: data.cpu
            )

            let jsonData = try JSONEncoder().encode(performanceFlutterData)
            var dict =
                try JSONSerialization.jsonObject(with: jsonData, options: .allowFragments)
                as? [String: Any]
            if dict == nil {
                throw NSError(
                    domain: "DictError", code: 0,
                    userInfo: [NSLocalizedDescriptionKey: "Dict is nil"])
            }
            dict?["eventChannelId"] = "performance_metrics"

            DispatchQueue.main.async {
                MedalliaDxaPlugin.flutterEventSink?(dict)
            }
        } catch {
            print("Performance JSON serialization error: \(error)")
        }
    }

    private func translateDeviceStressedTypeToString(
        nativeStresstedType value: DeviceStressedTypeFlutter
    ) -> String {
        var nativeStresstedType: String

        switch value {
        case .battery:
            nativeStresstedType = "battery"
        case .cpu:
            nativeStresstedType = "cpu"
        case .memory:
            nativeStresstedType = "memory"
        default:
            nativeStresstedType = "none"
        }
        return nativeStresstedType

    }

    private func translateTouchType(
        value: Int64
    ) -> FlutterTouchType {
        var type: FlutterTouchType

        switch value {
        case 0:
            type = FlutterTouchType.touchBegan
        case 1:
            type = FlutterTouchType.touchMoved
        case 2:
            type = FlutterTouchType.touchEnded
        default:
            type = FlutterTouchType.touchEnded
        }
        return type

    }

    func convertInt64ToInt(_ value: Int64?) -> Int? {
        guard let value = value else {
            return nil
        }
        return Int(value)
    }

}
final class PigeonError: Error {
    let code: String
    let message: String?
    let details: Any?

    init(code: String, message: String?, details: Any?) {
        self.code = code
        self.message = message
        self.details = details
    }

    var localizedDescription: String {
        return
            "PigeonError(code: \(code), message: \(message ?? "<nil>"), details: \(details ?? "<nil>")"
    }
}
